File: arrays/dutch_flag.py
import sys
import random
import itertools


def group_integers_inplace(numbers, pidx):
    """
    Groups (inplace) numbers by less-than-numbers[pidx],
    equal-to-numbers[pidx], and greater-than-numbers[pidx]

    Args:
        numbers (list[int]): list of integers
        pidx (int): index of pivot
    
    Returns:
        N/A
    """
    def swap(array, i, j):
        array[i], array[j] = array[j], array[i]

    pivot = numbers[pidx]
    low = 0
    # group integers that are less than the pivot
    for i in range(len(numbers)):
        if numbers[i] < pivot:
            swap(numbers, i, low)
            low += 1

    high = len(numbers) - 1
    # group integers that are greater than the pivot
    for i in range(high, low-1, -1):
        if numbers[i] > pivot:
            swap(numbers, i, high)
            high -= 1


def group_integers(numbers, pivot):
    """
    Returns a list where numbers are ordered
    by less-than-pivot, equal-to-pivot, and 
    greater-than-pivot.

    Args:
        numbers (list[int]): list of integers
        pivot (int): integer to order/group integers by

    Returns:
        ordered (list[int]): new list where numbers are
            ordered according to constraints listed in the 
            docstring
    """
    lt, eq, gt = [], [], []
    for number in numbers:
        if number < pivot:
            lt.append(number)
        elif number == pivot:
            eq.append(number)
        else:
            gt.append(number)
    return list(itertools.chain(lt, eq, gt))


def get_pivot_lo_hi_index(numbers, pivot):
    """
    Returns index of first and last pivot 
    occurrence 

    Args:
        numbers (list[int]): list of integers
        pivot (int): pivot value

    Returns:
        (lo, hi): indices of first and last
            pivot value occurrences
    """
    lo, hi = None, None
    for i, value in enumerate(numbers):
        if value == pivot:
            if lo is None:
                lo, hi = i, i
            else:
                hi += 1
    return lo, hi


def all_pivots_contiguous(numbers, pivot):
    """
    Returns true if all pivot values are in a contiguous 
    block in the list

    Args:
        numbers (list[int]): list of integers
        pivot (int): pivot value
    
    Returns:
        bool: True/False if block of 
            pivot values is/is-not conitguous
    """
    lo, hi = get_pivot_lo_hi_index(numbers, pivot)
    return all(v == pivot for v in numbers[lo: hi+1])


def all_numbers_grouped_around_pivot(numbers, pivot):
    """
    Returns true if all values less than the pivot 
    appear before the pivot in the array and all values
    greater than the pivot appear after the pivot in the 
    array

    Args:
        numbers (list[int]): list of integers
        pivot (int): pivot value 
        lo (int): index in ordered list where first
            occurrence of pivot appears
        hi (int): index in ordered list where last
            occurrence of pivot appears
    """
    lo, hi = get_pivot_lo_hi_index(numbers, pivot)
    lower = all(v < pivot for v in numbers[:lo])
    higher = all(v > pivot for v in numbers[hi+1:])
    return lower and higher


if __name__ == "__main__":
    size = int(sys.argv[1])
    lo, hi = 0, 25
    numbers = [random.randrange(lo, hi) for _ in range(size)]
    pidx = random.randrange(0, len(numbers))
    pivot = numbers[pidx]
    print("Original List:")
    print(numbers)
    print("Pivot: {}".format(pivot))
    print()
    print("Ordered List:")
    new_numbers = group_integers(numbers, pivot)
    print(group_integers(numbers, pivot))
    print()

    assert all_pivots_contiguous(new_numbers, pivot) == True
    assert all_numbers_grouped_around_pivot(new_numbers, pivot) == True
    print("All Test Cases Passed!")
    print()

    print("Original List:")
    print(numbers)
    print("Pivot: {}".format(pivot))
    print()
    print("Ordered List:")
    group_integers_inplace(numbers, pidx)
    print(numbers)
    print()

    assert all_pivots_contiguous(numbers, pivot) == True
    assert all_numbers_grouped_around_pivot(numbers, pivot) == True
    print("All Test Cases Passed!")
    

File: arrays/DutchFlag.java
import java.util.concurrent.ThreadLocalRandom;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

public class DutchFlag {

    public static int[] groupByPivot(int[] numbers, int pivot) {
        int index = 0, i;
        int[] ordered = new int[numbers.length];
        for (i = 0; i < numbers.length; i++) {
            if (numbers[i] < pivot) {
                ordered[index] = numbers[i];
                index++;
            }
        }

        for (i = 0; i < numbers.length; i++) {
            if (numbers[i] == pivot) {
                ordered[index] = numbers[i];
                index++;
            }
        }

        for (i = 0; i < numbers.length; i++) {
            if (numbers[i] > pivot) {
                ordered[index] = numbers[i];
                index++;
            }
        }
        return ordered;
    }

    public static void groupByPivotInplace(int [] numbers, int pidx) {
        int pivot = numbers[pidx];
        int low = 0;
        for (int i = 0; i < numbers.length; i++) {
            if (numbers[i] < pivot) {
                swap(numbers, i, low);
                low++;
            }
        }

        int high = numbers.length-1;
        for (int i = high; i >= low; i--) {
            if (numbers[i] > pivot) {
                swap(numbers, i, high);
                high--;
            }
        }
    }

    public static void swap(int [] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public static int[] getPivotSlicePosition(int[] numbers, int pivot) {
        int[] pos = new int[2];
        int found = 0;
        for (int i = 0; i < numbers.length; i++) {
            if (numbers[i] == pivot) {
                if (found == 0) {
                    pos[0] = i;
                    pos[1] = i;
                    found = 1;
                } else {
                    pos[1]++;
                }
            }
        }
        return pos;
    }

    public static boolean pivotSliceIsContiguous(int[] numbers, int pivot) {
        int[] slice = getPivotSlicePosition(numbers, pivot);
        for (int i = slice[0]; i <= slice[1]; i++) {
            if (numbers[i] != pivot) {
                return false;
            }
        }
        return true;
    }

    public static boolean numbersGroupedAroundPivot(int[] numbers, int pivot) {
        int[] slice = getPivotSlicePosition(numbers, pivot);
        for (int i = 0; i < slice[0]; i++) {
            if (numbers[i] >= pivot) {
                return false;
            }
        }

        for (int i = slice[1]+1; i < numbers.length; i++) {
            if (numbers[i] <= pivot) {
                return false;
            }
        }
        return true;
    }


    public static void main(String [] args) {
        int n = Integer.parseInt(args[0]);
        int[] numbers = new int[n];
        int min = 0, max = 25;
        for (int i = 0; i < n; i++) {
            numbers[i] = ThreadLocalRandom.current().nextInt(min, max+1);
        }

        // get pivot index and pivot
        int pidx = ThreadLocalRandom.current().nextInt(0, n);
        int pivot = numbers[pidx];

        System.out.println("original array:");
        System.out.println(Arrays.toString(numbers));
        System.out.println("pivot index: " + pidx);
        System.out.println("pivot: " + pivot);
        System.out.println();
        System.out.println("new array:");
        int[] ordered = groupByPivot(numbers, pivot);
        System.out.println(Arrays.toString(ordered));
        System.out.println();
        if (pivotSliceIsContiguous(ordered, pivot) && numbersGroupedAroundPivot(ordered, pivot)) {
            System.out.println("Tests Passed");
        } else {
            System.out.println("Tests Failed");
        }

        System.out.println();
        System.out.println("original array:");
        groupByPivotInplace(numbers, pidx);
        System.out.println("new array");
        System.out.println(Arrays.toString(numbers));
        System.out.println();
        if (pivotSliceIsContiguous(numbers, pivot) && numbersGroupedAroundPivot(numbers, pivot)) {
            System.out.println("Tests Passed");
        } else {
            System.out.println("Tests Failed");
        }

    }
}